spago.lock

```diff
@@ -2,840 +2,977 @@ workspace:
-        - string-parsers
-        - strings
-        - tailrec
-        - these
-        - transformers
-        - tuples
-        - type-equality
-        - typelevel-prelude
-        - unfoldable
-        - unicode
-        - unsafe-coerce
-        - variant
-        - versions
+      core:
+        dependencies:
+          - aff
+          - argonaut-core
+          - arrays
+          - codec-json
+          - console
+          - docs-search-common
+          - effect
+          - either
+          - foldable-traversable
+          - identity
+          - json
+          - json-codecs
+          - language-purescript
+          - lists
+          - maybe
+          - newtype
+          - node-buffer
+          - node-fs
+          - node-path
+          - node-process
+          - ordered-collections
+          - prelude
+          - profunctor
+          - search-trie
+          - string-parsers
+          - strings
+          - tuples
+          - unsafe-coerce
+        build_plan:
+          - aff
+          - aff-promise
+          - argonaut-core
+          - arraybuffer-types
+          - arrays
+          - bifunctors
+          - codec
+          - codec-json
+          - console
+          - const
+          - contravariant
+          - control
+          - datetime
+          - distributive
+          - docs-search-common
+          - effect
+          - either
+          - enums
+          - exceptions
+          - exists
+          - fixed-points
+          - foldable-traversable
+          - foreign
+          - foreign-object
+          - formatters
+          - functions
+          - functors
+          - gen
+          - identity
+          - integers
+          - invariant
+          - js-date
+          - json
+          - json-codecs
+          - language-purescript
+          - lazy
+          - lcg
+          - lists
+          - maybe
+          - newtype
+          - node-buffer
+          - node-event-emitter
+          - node-fs
+          - node-path
+          - node-process
+          - node-streams
+          - nonempty
+          - now
+          - nullable
+          - numbers
+          - ordered-collections
+          - orders
+          - parallel
+          - parsing
+          - partial
+          - posix-types
+          - prelude
+          - profunctor
+          - profunctor-lenses
+          - quickcheck
+          - quickcheck-laws
+          - random
+          - record
+          - refs
+          - safe-coerce
+          - search-trie
+          - st
+          - string-parsers
+          - strings
+          - tailrec
+          - these
+          - transformers
+          - tuples
+          - type-equality
+          - typelevel-prelude
+          - unfoldable
+          - unicode
+          - unsafe-coerce
+          - variant
+          - versions
+      test:
+        dependencies:
+          - exceptions
+          - spec
+          - spec-node
+        build_plan:
+          - aff
```

Comment:
The intended semantics is that `test.build_plan` does contain packages that are "additional" to the ones in `core.build_plan`. Sometimes it can happen that the test and core build plans intersect. This is because we're not doing an explicit diff, we're just building transitive closures separately - first starting with core dependencies, and then starting with test dependencies. If the test transitive closure also happens to touch some dependencies from core - so what, no big deal.

If you really insist, I can certainly add an explicit diff right before writing to the lockfile. This would be cleaner, but would incur additional code complexity and CPU cycles.

Reply:
Oh I see, we are already doing what I describe, minus the deduplication. I guess my concern is that the more lines there are in the yaml, the more it takes to parse it (see #1262); let's leave this as is, and sort out the yaml parsing performance otherwise